File: LowLevelDesign/task-scheduler-nodejs/src/ScheduledTask.js
import { randomUUID } from 'crypto';

/**
 * ScheduledTask - Wraps a Task with scheduling information
 * Implements comparison for priority queue ordering
 */
export class ScheduledTask {
    constructor(task, strategy) {
        this.id = randomUUID();
        this.task = task;
        this.strategy = strategy;
        this.nextExecutionTime = null;
        this.lastExecutionTime = null;
        this.updateNextExecutionTime();
    }

    updateNextExecutionTime() {
        this.nextExecutionTime = this.strategy.getNextExecutionTime(this.lastExecutionTime);
    }

    updateLastExecutionTime() {
        this.lastExecutionTime = this.nextExecutionTime;
    }

    getNextExecutionTime() {
        return this.nextExecutionTime;
    }

    getTask() {
        return this.task;
    }

    getId() {
        return this.id;
    }

    hasMoreExecutions() {
        return this.nextExecutionTime !== null && this.nextExecutionTime !== undefined;
    }

    // Comparison method for priority queue (earlier times have higher priority)
    compareTo(other) {
        if (!this.nextExecutionTime && !other.nextExecutionTime) return 0;
        if (!this.nextExecutionTime) return 1;
        if (!other.nextExecutionTime) return -1;
        return this.nextExecutionTime.getTime() - other.nextExecutionTime.getTime();
    }
}

File: SystemDesignDiagrams/practice/LLD/task-scheduler/src/main/java/com/taskscheduler/ScheduledTask.java
package com.taskscheduler;

import com.taskscheduler.strategy.SchedulingStrategy;
import com.taskscheduler.task.Task;

import java.time.LocalDateTime;
import java.util.Optional;
import java.util.UUID;

// implements Comparable
public class ScheduledTask implements Comparable<ScheduledTask> {
    private String id;
    private LocalDateTime nextExecutionTime;
    private LocalDateTime lastExecutionTime;
    private Task task;
    private SchedulingStrategy schedulingStrategy;

    public ScheduledTask(Task task, SchedulingStrategy strategy) {
        this.id = UUID.randomUUID().toString();
        this.task = task;
        this.schedulingStrategy = strategy;
        updateNextExecutionTime();
    }

    public String getId() {
        return this.id;
    }

    // This also
    public void updateNextExecutionTime() {
        Optional<LocalDateTime> nextTime = this.schedulingStrategy.execute(this.lastExecutionTime);
        this.nextExecutionTime = nextTime.orElse(null);
    }

    public LocalDateTime getNextExecutionTime() {
        return this.nextExecutionTime;
    }

    public Task getTask() {
        return this.task;
    }

    public void updateLastExecutionTime() {
        this.lastExecutionTime = nextExecutionTime;
    }

    // Signature of this method
    @Override
    public int compareTo(ScheduledTask o) {
        return this.nextExecutionTime.compareTo(o.nextExecutionTime);
    }

    public boolean hasMoreExecutions() {
        return this.nextExecutionTime != null;
    }
}
